// prettier-plugin-pug/dist/index.js
"use strict";
exports.__esModule = true;
var prettier_1 = require("prettier");
var lex = require("pug-lexer");
var logger_1 = require("./logger");
var makeString = prettier_1.util.makeString;
var logger = logger_1.createLogger(console);
if (process.env.NODE_ENV === 'test') {
    logger.setLogLevel(logger_1.LogLevel.DEBUG);
}
function quotationType(code) {
    var indexOfSingleQuote = code.indexOf("'");
    var indexOfDoubleQuote = code.indexOf('"');
    logger.debug({ code: code, indexOfSingleQuote: indexOfSingleQuote, indexOfDoubleQuote: indexOfDoubleQuote });
    if (indexOfSingleQuote === -1 && indexOfDoubleQuote === -1) {
        return undefined;
    }
    else if (indexOfSingleQuote === -1 && indexOfDoubleQuote !== -1) {
        return 'DOUBLE';
    }
    else if (indexOfDoubleQuote === -1 && indexOfSingleQuote !== -1) {
        return 'SINGLE';
    }
    else if (indexOfSingleQuote < indexOfDoubleQuote) {
        return 'SINGLE';
    }
    else if (indexOfDoubleQuote < indexOfSingleQuote) {
        return 'DOUBLE';
    }
    return;
}
function previousNormalAttributeToken(tokens, index) {
    for (var i = index - 1; i > 0; i--) {
        var token = tokens[i];
        if (token.type === 'start-attributes') {
            return;
        }
        if (token.type === 'attribute') {
            if (token.name !== 'class' && token.name !== 'id') {
                return token;
            }
        }
    }
    return;
}
exports.plugin = {
    languages: [
        {
            name: 'Pug',
            parsers: ['pug'],
            tmScope: 'text.jade',
            aceMode: 'jade',
            codemirrorMode: 'pug',
            codemirrorMimeType: 'text/x-pug',
            extensions: ['.jade', '.pug'],
            linguistLanguageId: 179,
            vscodeLanguageIds: ['jade']
        }
    ],
    parsers: {
        pug: {
            parse: function (text, parsers, options) {
                logger.debug('[parsers:pug:parse]:', { text: text });
                var tokens = lex(text, {});
                return tokens;
            },
            astFormat: 'pug-ast',
            hasPragma: function (text) {
                return text.startsWith('//- @prettier\n') || text.startsWith('//- @format\n');
            },
            locStart: function (node) {
                logger.debug('[parsers:pug:locStart]:', { node: node });
                return 0;
            },
            locEnd: function (node) {
                logger.debug('[parsers:pug:locEnd]:', { node: node });
                return 0;
            },
            preprocess: function (text, options) {
                logger.debug('[parsers:pug:preprocess]:', { text: text });
                return text;
            }
        }
    },
    printers: {
        'pug-ast': {
            print: function (path, _a, print) {
                var printWidth = _a.printWidth, singleQuote = _a.singleQuote, tabWidth = _a.tabWidth, useTabs = _a.useTabs;
                var tokens = path.stack[0];
                var result = '';
                var indentLevel = 0;
                var indent = ' '.repeat(tabWidth);
                if (useTabs) {
                    indent = '\t';
                }
                var pipelessText = false;
                var startTagPosition = 0;
                var startAttributePosition = 0;
                var previousAttributeRemapped = false;
                var wrapAttributes = false;
                for (var index = 0; index < tokens.length; index++) {
                    var token = tokens[index];
                    var previousToken = tokens[index - 1];
                    var nextToken = tokens[index + 1];
                    logger.debug('[printers:pug-ast:print]:', JSON.stringify(token));
                    switch (token.type) {
                        case 'tag':
                            if (previousToken) {
                                switch (previousToken.type) {
                                    case 'newline':
                                    case 'outdent':
                                        result += indent.repeat(indentLevel);
                                        break;
                                    case 'indent':
                                        result += indent;
                                        break;
                                }
                            }
                            if (!(token.val === 'div' && (nextToken.type === 'class' || nextToken.type === 'id'))) {
                                result += token.val;
                            }
                            startTagPosition = result.length;
                            break;
                        case 'start-attributes':
                            if (nextToken && nextToken.type === 'attribute') {
                                previousAttributeRemapped = false;
                                startAttributePosition = result.length;
                                result += '(';
                                var start = result.lastIndexOf('\n') + 1;
                                var lineLength = result.substring(start).length;
                                logger.debug(lineLength, printWidth);
                                var tempToken = nextToken;
                                var tempIndex = index + 1;
                                while (tempToken.type === 'attribute') {
                                    lineLength += tempToken.name.length + 1 + tempToken.val.toString().length;
                                    logger.debug(lineLength, printWidth);
                                    tempToken = tokens[++tempIndex];
                                }
                                if (lineLength > printWidth) {
                                    wrapAttributes = true;
                                }
                            }
                            break;
                        case 'attribute':
                            if (token.name === 'class' &&
                                typeof token.val === 'string' &&
                                (token.val.startsWith('"') || token.val.startsWith("'"))) {
                                var val_1 = token.val;
                                val_1 = val_1.substring(1, val_1.length - 1);
                                val_1 = val_1.trim();
                                val_1 = val_1.replace(/\s\s+/g, ' ');
                                var classes = val_1.split(' ');
                                var specialClasses = [];
                                var validClassNameRegex = /^-?[_a-zA-Z]+[_a-zA-Z0-9-]*$/;
                                for (var _i = 0, classes_1 = classes; _i < classes_1.length; _i++) {
                                    var className = classes_1[_i];
                                    if (!validClassNameRegex.test(className)) {
                                        specialClasses.push(className);
                                        continue;
                                    }
                                    var position_1 = startAttributePosition;
                                    result = [result.slice(0, position_1), "." + className, result.slice(position_1)].join('');
                                    startAttributePosition += 1 + className.length;
                                }
                                if (specialClasses.length > 0) {
                                    token.val = makeString(specialClasses.join(' '), singleQuote ? "'" : '"', false);
                                    previousAttributeRemapped = false;
                                }
                                else {
                                    previousAttributeRemapped = true;
                                    break;
                                }
                            }
                            else if (token.name === 'id' &&
                                typeof token.val === 'string' &&
                                (token.val.startsWith('"') || token.val.startsWith("'"))) {
                                var val_2 = token.val;
                                val_2 = val_2.substring(1, val_2.length - 1);
                                val_2 = val_2.trim();
                                var position_2 = startTagPosition;
                                result = [result.slice(0, position_2), "#" + val_2, result.slice(position_2)].join('');
                                startAttributePosition += 1 + val_2.length;
                                result = result.replace(/div#/, '#');
                                if (previousToken.type === 'attribute' && previousToken.name !== 'class') {
                                    previousAttributeRemapped = true;
                                }
                                break;
                            }
                            var hasNormalPreviousToken = previousNormalAttributeToken(tokens, index);
                            if (previousToken &&
                                previousToken.type === 'attribute' &&
                                (!previousAttributeRemapped || hasNormalPreviousToken)) {
                                result += ',';
                                if (!wrapAttributes) {
                                    result += ' ';
                                }
                            }
                            previousAttributeRemapped = false;
                            if (wrapAttributes) {
                                result += '\n';
                                result += indent.repeat(indentLevel + 1);
                            }
                            result += "" + token.name;
                            if (typeof token.val === 'boolean') {
                                if (token.val !== true) {
                                    result += "=" + token.val;
                                }
                            }
                            else {
                                var val_3 = token.val;
                                if (token.name.startsWith(':') || token.name.startsWith('v-bind:')) {
                                    val_3 = val_3.trim();
                                    val_3 = val_3.replace(/\s\s+/g, ' ');
                                    val_3 = val_3.replace('[ {', '[{').replace('} ]', '}]');
                                    val_3 = val_3.replace('[ (', '[(').replace(') ]', ')]');
                                    val_3 = val_3.replace('[ ', '[').replace(' ]', ']');
                                    var type = quotationType(val_3);
                                    if ((type === 'SINGLE' && !singleQuote) || (type === 'DOUBLE' && singleQuote)) {
                                        val_3 = val_3.replace(/['"]/g, function (match) { return (match === '"' ? "'" : '"'); });
                                    }
                                }
                                else if (/^["'](.*)["']$/.test(val_3)) {
                                    val_3 = makeString(val_3.slice(1, -1), singleQuote ? "'" : '"', false);
                                }
                                else if (val_3 === 'true') {
                                    break;
                                }
                                else {
                                    val_3 = val_3.trim().replace(/\s\s+/g, ' ');
                                    if (val_3.startsWith('{ ')) {
                                        val_3 = "{" + val_3.substring(2, val_3.length);
                                    }
                                }
                                result += "=" + val_3;
                            }
                            break;
                        case 'end-attributes':
                            if (wrapAttributes) {
                                result += '\n';
                                result += indent.repeat(indentLevel);
                            }
                            wrapAttributes = false;
                            if (result.endsWith('(')) {
                                result = result.substring(0, result.length - 1);
                            }
                            else if (previousToken && previousToken.type === 'attribute') {
                                result += ')';
                            }
                            if (nextToken && (nextToken.type === 'text' || nextToken.type === 'path')) {
                                result += ' ';
                            }
                            break;
                        case 'indent':
                            result += '\n';
                            result += indent.repeat(indentLevel);
                            indentLevel++;
                            break;
                        case 'outdent':
                            if (previousToken && previousToken.type !== 'outdent') {
                                if (token.loc.start.line - previousToken.loc.end.line > 1) {
                                    result += '\n';
                                }
                                result += '\n';
                            }
                            indentLevel--;
                            break;
                        case 'class':
                            if (previousToken) {
                                switch (previousToken.type) {
                                    case 'newline':
                                    case 'outdent':
                                        result += indent.repeat(indentLevel);
                                        break;
                                    case 'indent':
                                        result += indent;
                                        break;
                                }
                            }
                            result += "." + token.val;
                            if (nextToken && nextToken.type === 'text') {
                                result += ' ';
                            }
                            break;
                        case 'eos':
                            while (result.endsWith('\n')) {
                                result = result.substring(0, result.length - 1);
                            }
                            result += '\n';
                            break;
                        case 'comment':
                            if (previousToken) {
                                switch (previousToken.type) {
                                    case 'newline':
                                    case 'outdent':
                                        result += indent.repeat(indentLevel);
                                        break;
                                    case 'indent':
                                        result += indent;
                                        break;
                                }
                            }
                            result += "//" + (token.buffer ? '' : '-') + token.val.replace(/\s\s+/g, ' ');
                            break;
                        case 'newline':
                            if (previousToken && token.loc.start.line - previousToken.loc.end.line > 1) {
                                result += '\n';
                            }
                            result += '\n';
                            break;
                        case 'text':
                            var val = token.val;
                            val = val.replace(/\s\s+/g, ' ');
                            if (previousToken) {
                                switch (previousToken.type) {
                                    case 'newline':
                                        if (pipelessText === false) {
                                            result += indent.repeat(indentLevel);
                                            if (/^ .+$/.test(val)) {
                                                result += '|\n';
                                                result += indent.repeat(indentLevel);
                                            }
                                            result += '|';
                                            if (/.*\S.*/.test(token.val)) {
                                                result += ' ';
                                            }
                                        }
                                        else {
                                            result += indent.repeat(indentLevel);
                                            result += indent;
                                        }
                                        break;
                                    case 'indent':
                                        result += indent;
                                        result += '|';
                                        if (/.*\S.*/.test(token.val)) {
                                            result += ' ';
                                        }
                                        break;
                                    case 'start-pipeless-text':
                                        result += indent;
                                        break;
                                    case 'interpolated-code':
                                        if (/^ .+$/.test(val)) {
                                            result += ' ';
                                        }
                                        break;
                                }
                            }
                            var needsTrailingWhitespace = false;
                            if (nextToken && nextToken.type === 'interpolated-code' && val.endsWith(' ')) {
                                needsTrailingWhitespace = true;
                            }
                            val = val.trim();
                            if (val.startsWith('{{') && val.endsWith('}}')) {
                                var code = val.substring(2, val.length - 2);
                                code = code.trim();
                                var type = quotationType(code);
                                if ((type === 'SINGLE' && singleQuote) || (type === 'DOUBLE' && !singleQuote)) {
                                    val = '{{ ';
                                    val += code.replace(/['"]/g, function (match) { return (match === '"' ? "'" : '"'); });
                                    val += ' }}';
                                }
                            }
                            if (previousToken && (previousToken.type === 'tag' || previousToken.type === 'id')) {
                                val = " " + val;
                            }
                            result += val;
                            if (needsTrailingWhitespace) {
                                result += ' ';
                            }
                            break;
                        case 'interpolated-code':
                            if (previousToken && previousToken.type === 'tag') {
                                result += ' ';
                            }
                            result += "#{" + token.val + "}";
                            break;
                        case 'code':
                            if (indentLevel > 0) {
                                result += indent;
                            }
                            result += "- " + token.val;
                            break;
                        case 'id':
                            var lastPositionOfNewline = result.lastIndexOf('\n');
                            if (lastPositionOfNewline === -1) {
                                lastPositionOfNewline = 0;
                            }
                            var position = result.indexOf('.', lastPositionOfNewline);
                            if (position === -1) {
                                position = result.length;
                            }
                            var _indent = '';
                            if (previousToken) {
                                switch (previousToken.type) {
                                    case 'newline':
                                    case 'outdent':
                                        _indent = indent.repeat(indentLevel);
                                        break;
                                    case 'indent':
                                        _indent = indent;
                                        break;
                                }
                            }
                            result = [result.slice(0, position), _indent, "#" + token.val, result.slice(position)].join('');
                            break;
                        case 'start-pipeless-text':
                            pipelessText = true;
                            result += '\n';
                            result += indent.repeat(indentLevel);
                            break;
                        case 'end-pipeless-text':
                            pipelessText = false;
                            break;
                        case 'doctype':
                            result += "doctype " + token.val;
                            break;
                        case 'dot':
                            result += '.';
                            break;
                        case 'block':
                            if (previousToken) {
                                switch (previousToken.type) {
                                    case 'newline':
                                    case 'outdent':
                                        result += indent.repeat(indentLevel);
                                        break;
                                    case 'indent':
                                        result += indent;
                                        break;
                                }
                            }
                            result += 'block ';
                            if (token.mode !== 'replace') {
                                result += token.mode;
                            }
                            result += token.val;
                            break;
                        case 'extends':
                            result += 'extends ';
                            break;
                        case 'path':
                            result += token.val;
                            break;
                        case 'start-pug-interpolation':
                            result += '#[';
                            break;
                        case 'end-pug-interpolation':
                            result += ']';
                            break;
                        case 'include':
                            if (previousToken && previousToken.type === 'indent') {
                                result += indent;
                            }
                            result += 'include';
                            break;
                        case 'filter':
                            result += ":" + token.val;
                            break;
                        default:
                            throw new Error('Unhandled token: ' + JSON.stringify(token));
                    }
                }
                logger.debug(result);
                return result;
            },
            embed: function (path, print, textToDoc, options) {
                return null;
            },
            insertPragma: function (text) {
                return "//- @prettier\n" + text;
            }
        }
    },
    options: [],
    defaultOptions: {}
};
exports.languages = exports.plugin.languages;
exports.parsers = exports.plugin.parsers;
exports.printers = exports.plugin.printers;
exports.options = exports.plugin.options;
exports.defaultOptions = exports.plugin.defaultOptions;
