// prettier-plugin-pug/dist/logger.js
"use strict";
exports.__esModule = true;
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["DEBUG"] = 0] = "DEBUG";
    LogLevel[LogLevel["LOG"] = 1] = "LOG";
    LogLevel[LogLevel["INFO"] = 2] = "INFO";
    LogLevel[LogLevel["WARN"] = 3] = "WARN";
    LogLevel[LogLevel["ERROR"] = 4] = "ERROR";
    LogLevel[LogLevel["OFF"] = 5] = "OFF";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
var Logger = (function () {
    function Logger(logger, level) {
        if (logger === void 0) { logger = console; }
        if (level === void 0) { level = LogLevel.INFO; }
        this.logger = logger;
        this.level = level;
    }
    Logger.prototype.setLogLevel = function (level) {
        this.level = level;
    };
    Logger.prototype.debug = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        this.message.apply(this, [LogLevel.DEBUG, message].concat(optionalParams));
    };
    Logger.prototype.log = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        this.message.apply(this, [LogLevel.LOG, message].concat(optionalParams));
    };
    Logger.prototype.info = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        this.message.apply(this, [LogLevel.INFO, message].concat(optionalParams));
    };
    Logger.prototype.warn = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        this.message.apply(this, [LogLevel.WARN, message].concat(optionalParams));
    };
    Logger.prototype.error = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        this.message.apply(this, [LogLevel.ERROR, message].concat(optionalParams));
    };
    Logger.prototype.message = function (level, message) {
        var _a;
        var optionalParams = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            optionalParams[_i - 2] = arguments[_i];
        }
        if (this.level !== LogLevel.OFF && this.level <= level) {
            (_a = this.logger)[Logger.LOG_LEVELS[level]].apply(_a, [message].concat(optionalParams));
        }
    };
    Logger.LOG_LEVELS = [
        'debug',
        'log',
        'info',
        'warn',
        'error'
    ];
    return Logger;
}());
exports.Logger = Logger;
function createLogger(logger) {
    if (logger === void 0) { logger = console; }
    return new Logger(logger);
}
exports.createLogger = createLogger;
